### PR TITLE
Add StateList

### PR DESCRIPTION
**Pull request by micahkemp**
This PR adds the StateList command.

### CODE REVIEW

File: tfexec/options.go [hunk 1]
```diff
@@ -4,6 +4,16 @@ import (
 	"encoding/json"
 )
 
+// AddressOption represents an address.
```
Comment (micahkemp) [inline tfexec/options.go:7]: There may be opinions regarding how this is named, or even implemented. `terraform state show` is documented as taking zero or more addresses as options.  As such I added an AddressOption, which ends up getting applied (for `StateList`) as positional arguments after all the other options.

I'm more than happy to change how this is implemented if there are other suggestions.

**Review by radeksimko — COMMENTED** [review]
Thank you for this PR as well.

I have the same question for this command as in your [other PR](https://github.com/hashicorp/terraform-exec/pull/205):
Could you share your use case for this command in the context of automation?

Keep in mind that this command (unlike most others we integrated) does not provide machine-readable (JSON) output and AFAIK therefore doesn't provide any compatibility guarantees for the output format.

If we have a use case though we could propose enhancement of Terraform CLI so that it does offer JSON output - this would then make the integration much easier here.

**Comment by micahkemp** [comment]
Use case described [here](https://github.com/hashicorp/terraform-exec/pull/205#issuecomment-896089101) on #205.

I'll also mention that as these two PRs are related in terms of how I would use them, and because they're both being submitted at the same time, I'm more than happy to combine them into one.  I kept them split to hopefully make them simpler to review and also to allow you flexibility in reviewing them as separate concepts.

**Comment by micahkemp** [comment]
As discussed in the comments of #205, I think it makes sense for wait until `state show` supports `-json`.

**Comment by radeksimko** [comment]
Yep, I will be doing some exploratory work which will likely result adding JSON output for more commands and I will certainly use your use case in the process to advocate for `state list -json` specifically.

I will let you know here once any relevant PR is ready upstream.

Thanks.

**Comment by radeksimko** [comment]
As mentioned in https://github.com/hashicorp/terraform-exec/pull/205#issuecomment-1172203797 the use case can also be satisfied with existing functionality. See below.

```sh
$ terraform state list
random_pet.adam
random_pet.basil
random_pet.claire
```
which is equivalent to
```sh
$ terraform show -json | jq -r '.values.root_module.resources[].address'
random_pet.adam
random_pet.basil
random_pet.claire
```
and similarly to
```go
package main

import (
	"context"
	"fmt"
	"log"
	"os"
	"os/exec"

	"github.com/hashicorp/terraform-exec/tfexec"
)

func main() {
	workDir, err := os.Getwd()
	if err != nil {
		log.Fatal(err)
	}

	execPath, err := exec.LookPath("terraform")
	if err != nil {
		log.Fatal(err)
	}

	tf, err := tfexec.NewTerraform(workDir, execPath)
	if err != nil {
		log.Fatal(err)
	}

	ctx := context.Background()
	state, err := tf.Show(ctx)
	if err != nil {
		log.Fatal(err)
	}

	for _, resource := range state.Values.RootModule.Resources {
		fmt.Println(resource.Address)
	}
}
```

I hope that helps.

**Pull request closed without merging**